add test for log::FileSink

kukinsula opened the pull request:
I managed to test log messages through a FileSink: messages are written in /tmp/openage-log-test (maybe I should change that file name because there's already /tmp/openage-log and it can get ambiguous). So to see the messages you just need to read that file (cat /tmp/openage-log-test).

I set the FileSink's loglevel to err to ignore messages with a lower loglevel and it works fine.

TheJJ commented:
Thanks!

A few comments:
* The commit email is not associated with github. Either update the email in the commit or register it to github please.
* You didn't add yourself to `copying.md`
* You do a good job writing to that filesink, but do not verify its expected contents.
* You could do more tests that will check wether only the assigned log levels are written to that file
* Please remove the file after testing, we don't wanna clutter the filesystem :wink:

kukinsula commented:
Hi TheJJ !

I corrected what you told me to fix, I hope this is ok now.

To verify that only relevant messages are written in the log file I read that file (line by line) looking for "ERR" or "CRIT", is it ok or were you thinking some other way ?

I also added another test: I log messages through a FileSink with a crit log level, then read the log file which should be empty.

One last question, should I ensure that all unit tests don't return -1 to state wether it passed or failed (like other subsystems tests) ?

mic-e commented on cpp/log/test.cpp:
```diff
@@ -61,4 +68,76 @@ void test() {
+ * than level::err will be printed in that file.
+ */
+void file_log_sink () {
+	const char *filelog_name = "/tmp/openage-log-test";
```
hard-coding the filename to `/tmp` is sub-optimal; this might not work on OSX and will not work on Win32. Use `mkstemp` instead.

kukinsula replied:
Thank you, I'll take the time (soon) to change that.

kukinsula replied:
Hey @mic-e ! I am trying to use mkstemp to create a temporary file (it has a random name /tmp/file.XXXXXX) but I can't find a way to retrieve that random name so I can give it to the FileSink. What were you exactly thinking about mkstemp ? or maybe I am missing something.

detrumi replied:
There's an example on mkstemp usage [here](http://www.thegeekstuff.com/2012/06/c-temporary-files/).

mic-e replied:
Sorry, looks like `mstemp` is _not_ the function you're looking for; you'll want to use `tmpnam` instead.

I guess I got irritated by the man page's warning to never use `tmpnam`, and to use `mkstemp` instead. Which is bullshit in our case.

mic-e commented on cpp/log/test.cpp:
```diff
@@ -61,4 +68,76 @@ void test() {
+	logger.log(MSG(crit) << "This message should appear in openage-log-test");
+
+	std::string line;
+	std::ifstream infile(filelog_name);
```
I think you should delete the FileSink object before opening the file for reading - who knows what sort of buffering might be happening in the background.

Also, I think Win32 doesn't allow you to open a file that has already been opened for writing...

StepS- replied:
@mic-e it does allow opening the file but it needs to have the "share" access. Depending on how the file was opened for writing (with or without access sharing for reading or writing), the file may or may not be accessible for reading.

TheJJ replied:
If you don't want to close the file, you should at least flush the buffer to write out anything cached.

mic-e commented on cpp/log/test.cpp:
```diff
@@ -61,4 +68,76 @@ void test() {
+void file_log_sink () {
+	const char *filelog_name = "/tmp/openage-log-test";
+	TestLogSource logger;
+	FileSink filelog(filelog_name, true);
```
You'll want to pass `false` here. If you pass `true`, it will append to the existing logfile, which may lead to wrong results.

kukinsula replied:
Will do this too then (since I delete the file at the end of the test it doesn't lead to wrong results right now).

TheJJ commented on cpp/log/test.cpp:
```diff
@@ -59,6 +64,104 @@ void test() {
+	std::ifstream infile(filelog_name);
+
+	while (std::getline(infile, line)) {
+		if ((line.find("ERR") == std::string::npos) && (line.find("CRIT") == std::string::npos)) {
```
i'd recommend searching for `"should not appear"`, because this is the essential part to test, not the line prefix which may change any time.

TheJJ commented on cpp/log/test.cpp:
```diff
@@ -59,6 +64,104 @@ void test() {
+ * message into the the filelog because it
+ * has a too low log level.
+ */
+int empty_file_log_sink () {
```
i'm not sure whether this test does any new coverage, the do-we-output-for-some-level test is already done above.

TheJJ commented on cpp/log/test.cpp:
```diff
@@ -59,6 +64,104 @@ void test() {
+	char filelog_name[L_tmpnam];
+	int result = 0;
+
+	tmpnam(filelog_name);
```
You might want to use http://linux.die.net/man/3/mkstemp.
It returns the file descriptor directly, instead of just a name, which is way easier in handling and doesn't require the `filelog_name` variable.

TheJJ replied:
If you want/must stick to generating a filename only, you should use at least use [tempnam](http://linux.die.net/man/3/tempnam) instead.

TheJJ commented:
please update this test to the new project layout and testing system.

TheJJ commented:
If you wanna continue this, feel free to request a reopen.
